### DISCUSSION TITLE
Github Actions fix caching timeout

### DISCUSSION THREAD

**Discussion by bryaneaton13** [vercel/vercel#9340] [General] [answered]
## TL;DR
If restoring cache is 🐢 , 🛑 it and move on

There's a bug with [caching github actions](https://github.com/actions/cache/issues/810) where it could potentially spend **1 hour** trying to restore cache items (like pnpm cache for vercel). This could cause an action to take way longer than it should.

## Description
I'm following the [contributing guidelines](https://github.com/vercel/vercel/blob/main/.github/CONTRIBUTING.md) by discussing the change here first.

I see that you've implemented a timeout already, but that was in the past when you were using the built in caching with the `actions/setup-node` Github Action. That used a property called `timeout-minutes` as you can see implemented in [this commit](https://github.com/vercel/vercel/commit/baa56aed2cb13ca0ef412868c3c16336e0936f5d)

Since you've since switched to using your own `actions/cache` Github Action, you'll need to reimplement the cache timeout using an env variable with that action: `SEGMENT_DOWNLOAD_TIMEOUT_MINS: 5`

## How to fix
1. Find all the github actions that use `timeout-minutes: ` as part of the `actions/setup-node@v3` action
2. Remove that
3. Right below that action you'll find `uses: actions/cache@v3`
4. Set an env variable:

```sh
env:
  SEGMENT_DOWNLOAD_TIMEOUT_MINS: 5 # See https://github.com/actions/cache/issues/810
```

### Example

```diff
      - uses: actions/setup-node@v3
-        timeout-minutes: 5 # See https://github.com/actions/cache/issues/810
        with:
          node-version: ${{ matrix.node }}
      - uses: actions/cache@v3
        with:
          path: '**/node_modules'
          key: pnpm-${{ matrix.os }}-${{ matrix.node }}-${{ hashFiles('pnpm-lock.yaml') }}
          restore-keys: pnpm-${{ matrix.os }}-${{ matrix.node }}```
+      env:
+        SEGMENT_DOWNLOAD_TIMEOUT_MINS: 3 # Limit the time this can run
```

## Why am I doing this?
I just ran into this issue last week in my own monorepo when I was trying to optimize our github caching since we have a lot of github actions that need to cache and install deps. I saw this comment (https://github.com/actions/cache/issues/810#issuecomment-1328163060) about the same thing happening in vercel's repo.

**Comment by bryaneaton13** [comment, marked as answer]
Fixed in this PR - https://github.com/vercel/vercel/pull/9342

**Reply by amyegan** [reply]
Thanks for sharing this!
